File: cloudflare/src/requestid.ts
/**
 * Request ID generation and propagation.
 *
 * Mirrors the Go implementation in proxy/requestid.go:
 * - Generates a UUID v4 if no X-Request-ID header is provided.
 * - Preserves client-provided IDs.
 * - Echoes the ID back in the response header.
 */

export const REQUEST_ID_HEADER = "X-Request-ID";

/**
 * Generate a cryptographically random UUID v4.
 */
export function generateRequestID(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);

  // Set version (4) and variant (RFC 4122) bits.
  bytes[6] = (bytes[6] & 0x0f) | 0x40;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;

  const hex = Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

  return [
    hex.slice(0, 8),
    hex.slice(8, 12),
    hex.slice(12, 16),
    hex.slice(16, 20),
    hex.slice(20, 32),
  ].join("-");
}

/**
 * Extracts or generates the request ID from headers.
 */
export function getRequestID(headers: Headers): string {
  return headers.get(REQUEST_ID_HEADER) || generateRequestID();
}

File: proxy/requestid.go
package proxy

import (
	"context"
	"crypto/rand"
	"fmt"
	"net/http"
)

type contextKey string

const requestIDKey contextKey = "request-id"

// RequestIDHeader is the header name used to carry the request ID.
const RequestIDHeader = "X-Request-ID"

// generateRequestID creates a random UUID v4 string.
func generateRequestID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		panic("go-drp: crypto/rand unavailable: " + err.Error())
	}
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant bits
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}

// requestIDFromContext retrieves the request ID stored in the context.
func requestIDFromContext(ctx context.Context) string {
	if id, ok := ctx.Value(requestIDKey).(string); ok {
		return id
	}
	return ""
}

// requestIDMiddleware ensures every request carries an X-Request-ID. If the
// client supplies one it is preserved and echoed in the response header;
// otherwise a fresh UUID v4 is generated.
func requestIDMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		id := r.Header.Get(RequestIDHeader)
		if id == "" {
			id = generateRequestID()
			r.Header.Set(RequestIDHeader, id)
		}
		w.Header().Set(RequestIDHeader, id)
		ctx := context.WithValue(r.Context(), requestIDKey, id)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}
